docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+    dimension_ of the array_. The chunk shape is the tuple of interval
+    lengths, and the chunk size (i.e., number of elements_ contained
+    within the chunk) is the product of its interval lengths.
+
+    The chunk shape elements are non-zero when the corresponding dimensions of
+    the arrays are of non-zero length.
+
+.. _grid:
+.. _grids:
+
+*Grid*
+
+    The chunks_ of an array_ are organised into a grid. This
+    specification only considers the case where all chunks_ have the
+    same chunk shape and the chunks form a regular grid. However,
+    extensions may define other grid types such as
+    rectilinear grids.
+
+.. _memory layout:
+.. _memory layouts:
+
+*Memory layout*
+
+    An array_ is associated with a memory layout which defines how to
+    construct a binary representation of a single chunk_ by organising
+    the binary values of the elements_ within the chunk_ into a single
+    contiguous sequence of bytes. This specification defines two types
+    of memory layout based on "C" (row-major) and "F" (column-major)
+    ordering of elements_, but extensions may define other
+    memory layouts.
+
+.. _compressor:
+.. _compressors:
+
+*Compressor*
+
+    An array_ may be associated with a compressor, which is a codec_
+    that transforms the binary representation of a chunk_ in some way,
+    usually to reduce data size.
+
+.. _codec:
+.. _codecs:
+
+*Codec*
+
+    A codec is a pair of algorithms which transform binary data in
+    some way and are used to encode and decode chunks_. This
+    specification defines a codec interface which comprises
+    a pair of operations, one to perform the transformation (encode),
+    the other to reverse the transformation (decode). This
+    specification only considers the case where a codec is used as a
+    compressor_, but extensions may extend the chunk encoding
+    process, for example, to add support for one or more filter codecs
+    to be applied prior to compression.
+
+.. _metadata document:
+.. _metadata documents:
+
+*Metadata document*
+
+    Each array_ in a hierarchy_ is represented by a metadata document,
+    which is a machine-readable document containing essential
+    processing information about the node. For example, an array_
+    metadata document will specify the number of dimensions_, shape_,
+    `data type`_, grid_, `memory layout`_ and compressor_ for that
+    array_.
+
+    Groups can have a optional metadata document which provide extra
+    information about a group.
+
+.. _store:
+.. _stores:
+
+*Store*
+
+    The `metadata documents`_ and encoded chunk_ data for all nodes in
+    a hierarchy_ are held in a store. To enable a variety of different
+    store types to be used, this specification defines an `Abstract store
+    interface`_ which is a common set of operations that stores may
+    provide.
+
+.. _storage transformer:
+.. _storage transformers:
+
+*Storage transformer*
```

Comment:
This one is also quite unclear (haven't read the details yet, not sure if the goal is that the definition itself should make sense.)

Reply:
I think the definition should provide an intuitive idea what storage transformers are, I'd be happy do adopt. However, I'm not quite sure what it's lacking. The [description below in the storage transformers section](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#id19) is a bit more descriptive, does this give a better idea?

> A Zarr storage transformer allows to change the zarr-compatible data before storing it. The stored transformed data is restored to its original state whenever data is requested by the Array. Storage transformers can be configured per array via the [storage_transformers](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#id14) name in the [array metadata](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#id9). Storage transformers which do not change the storage layout (e.g. for caching) may be specified at runtime without adding them to the array metadata.

> A storage transformer serves the same [abstract store interface](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#id17) as the [store](https://zarr-specs.readthedocs.io/en/latest/core/v3.0.html#store). However, it should not persistently store any information necessary to restore the original data, but instead propagates this to the next storage transformer or the final store. From the perspective of an array or a previous stage transformer both store and storage transformer follow the same protocol and can be interchanged regarding the protocol. The behaviour can still be different, e.g. requests may be cached or the form of the underlying data can change.

----------------------------

> This seems like a generalization of a codec (similar to "filter" in v2)

I think storage transformers are to stores what filters are to codecs. It allows to have general properties outside of the store, but is different to codecs, as it rather handles the storage format itself, **after** chunking and indexing is done.

One concret example is [this sharding extension](https://zarr-specs.readthedocs.io/en/latest/extensions/storage-transformers/sharding/v1.0.html), which would not be possible in a codec or filter.

------------------------

Does this help to give a more clear idea? It seems like this definition is lacking some parts, but I'm not quite sure which ones, do you maybe have concrete suggestions, @constantinpape & @manzt?

PS: I also just openend a PR for it on the zarr-python implementation: https://github.com/zarr-developers/zarr-python/pull/1096, not sure if this provides more insights, though.

Reply:
Thanks for the clarification, @jstriebel. I think adding one less technical sentence here could help; for me this explanation
> It allows to have general properties outside of the store, but is different to codecs, as it rather handles the storage format itself, after chunking and indexing is done.

was quite useful, maybe you could add something along these lines?

Reply:
I tried to clarify the storage transformer section in https://github.com/alimanfoo/zarr-specs/pull/1 using and illustration and adding more explanation to the text. Does this help, would you like to see further changes? Please feel free to suggest changes, also directly in https://github.com/alimanfoo/zarr-specs/pull/1 if you want.
cc @constantinpape @manzt @rabernat

Reply:
Those changes have been merged in #163, marking as resolved for now.